File: built-in plugins dsp/mb-reverb-spring.cpp
/**
 * MB Spring Reverb
 * Category : effect
 * Type     : reverb
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Vintage spring reverb
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_REVERB_SPRING_H
#define MB_REVERB_SPRING_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbReverbSpring : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-reverb-spring";
    static constexpr const char* PLUGIN_NAME    = "MB Spring Reverb";
    static constexpr const char* PLUGIN_TYPE    = "reverb";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float tension = 0.5f;  // range [0, 1]
    float decay = 1.5f;  // range [0.3, 4]
    float mix = 0.3f;  // range [0, 1]
    };

    MbReverbSpring() = default;
    ~MbReverbSpring() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.tension = std::clamp(params.tension, 0f, 1f);
        params.decay = std::clamp(params.decay, 0.3f, 4f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Spring Reverb
        return input;
    }
};

#endif // MB_REVERB_SPRING_H

File: built-in plugins dsp/mb-reverb-spring.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbReverbSpringPlugin: PluginDefinition = { id: 'mb-reverb-spring', slug: 'mb-reverb-spring', name: 'MB Spring Reverb', category: 'effect', type: 'reverb', version: '1.0.0', description: 'Vintage spring reverb', author: 'Max Booster', grade: 'A', parameters: [{ id: 'tension', name: 'Tension', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'decay', name: 'Decay', type: 'float', defaultValue: 1.5, minValue: 0.3, maxValue: 4, automatable: true }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 0.3, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { tension: 0.5, decay: 1.5, mix: 0.3 } };

export default MbReverbSpringPlugin;
